pkg/datapath/linux/node.go

```diff
@@ -1368,6 +1368,50 @@ func (n *linuxNodeHandler) NodeNeighborRefresh(ctx context.Context, nodeToRefres
 	}
 }
 
+// NodeCleanNeighbors cleans all neighbor entries for the direct routing device
+// and the encrypt interface. It should be used when the agent changes the
+// state from `n.enableNeighDiscovery = true` to
+// `n.enableNeighDiscovery = false`.
+func (n *linuxNodeHandler) NodeCleanNeighbors() {
+	linksToClean := []string{option.Config.DirectRoutingDevice}
+	if option.Config.EncryptInterface != option.Config.DirectRoutingDevice {
+		linksToClean = append(linksToClean, option.Config.EncryptInterface)
+	}
```

Comment:
Did you check that the `option.Config.EncryptInterface` slice has been populated / devices autodetected at the time when this logic is called?

Reply:
For the `option.Config.EncryptInterface` it happens here https://github.com/cilium/cilium/blob/5cf3c130f8a38c9683115e6af3c258752907f25a/daemon/cmd/daemon_main.go#L1171 which is done in the `initEnv()` https://github.com/cilium/cilium/blob/5cf3c130f8a38c9683115e6af3c258752907f25a/daemon/cmd/daemon_main.go#L126-L128
The caller of `CleanNeigbors` is executed in the `runDaemon()` function.

For the `option.Config.DirectRoutingDevice`, it happens here:
https://github.com/cilium/cilium/blob/5cf3c130f8a38c9683115e6af3c258752907f25a/daemon/cmd/daemon_main.go#L1330-L1332

which is a couple lines above the `CleanNeigbors` call

Reply:
Oh, I guess this is simpler because this is based against v1.8 branch. Upcoming encryption backports will change this (to `pkg/datapath/loader/base.go`), so I guess we'll just have to keep a close eye out when those backports are proposed.

Reply:
^ Note, we have now merged encryption backports so might want to double-check these paths after the next rebase.

Reply:
Encryption side looks good to me before the interface store we have an encryption branch that sets ifaceName.

```
                case option.Config.EnableNodePort:
                        ifaceName = option.Config.DirectRoutingDevice
                        n.enableNeighDiscovery = true
                case n.nodeConfig.EnableIPSec &&
                        option.Config.Tunnel == option.TunnelDisabled &&
                        len(option.Config.EncryptInterface) != 0:
                        // When FIB lookup is not supported we need to pick an
                        // interface so pick first interface in the list. On
                        // kernels with FIB lookup helpers we do a lookup from
                        // the datapath side and ignore this value.
                        ifaceName = option.Config.EncryptInterface[0]
                        n.enableNeighDiscovery = true
                }
```
So should be the same as NodePort.

Reply:
And  EncryptInterface is autodiscovered from Reinitialize() called just before NodeConfigurationChanged so our autodiscovery should be working OK with this patch.

```
        if err := l.reinitializeIPSec(ctx); err != nil {
                return err
        }

        if err := o.Datapath().Node().NodeConfigurationChanged(*o.LocalConfig()); err != nil {
                return err
        }
```